File: static/credential-crypto.js
(function (root, factory) {
  if (typeof module !== 'undefined' && module.exports) {
    module.exports = factory();
  } else {
    root.ChatCredentialCrypto = factory();
  }
}(typeof globalThis !== 'undefined' ? globalThis : this, function () {
  'use strict';

  // Credential derivation/wrap/PRF algorithms shared by enc-key.js.
  // PBKDF2 derivation (100000 x SHA-256 -> 256 bits), AES-GCM wrap-key
  // generation/probe, data-key wrap/unwrap, authenticator PRF import/eval
  // bytes plus base64 transforms. Every browser capability arrives via
  // explicit env params (no ambient store, DOM, or key-flow reach);
  // enc-key.js passes browser globals at the call site and keeps store
  // lifecycle plus the EncKey surface.

  var PBKDF2_ITERATIONS = 100000;
  var PBKDF2_HASH = 'SHA-256';
  var PBKDF2_KEY_BITS = 256;
  var PRF_WRAP_LABEL = 'chatbot-enc-key-wrap-v1';

  function requireFn(env, name) {
    if (!env || typeof env[name] !== 'function') {
      throw new Error('ChatCredentialCrypto requires ' + name);
    }
    return env[name];
  }

  function requireSubtle(env) {
    if (!env || !env.subtle) {
      throw new Error('ChatCredentialCrypto requires subtle');
    }
    return env.subtle;
  }

  function requireTextEncoder(env) {
    if (!env || typeof env.TextEncoderImpl !== 'function') {
      throw new Error('ChatCredentialCrypto requires TextEncoderImpl');
    }
    return env.TextEncoderImpl;
  }

  // Base64 encode for raw key bytes (byte-wise String.fromCharCode shape).
  function encodeBase64(bytes, btoaImpl) {
    if (typeof btoaImpl !== 'function') {
      throw new Error('ChatCredentialCrypto requires btoaImpl');
    }
    var binary = '';
    bytes.forEach(function (b) {
      binary += String.fromCharCode(b);
    });
    return btoaImpl(binary);
  }

  // Base64 decode to bytes.
  function decodeBase64(value, atobImpl) {
    if (typeof atobImpl !== 'function') {
      throw new Error('ChatCredentialCrypto requires atobImpl');
    }
    var binary = atobImpl(value);
    var bytes = new Uint8Array(binary.length);
    for (var i = 0; i < binary.length; i += 1) {
      bytes[i] = binary.charCodeAt(i);
    }
    return bytes;
  }

  // Salt bytes for PBKDF2 from the server-provided base64 salt.
  function decodeSaltB64(saltB64, atobImpl) {
    if (typeof atobImpl !== 'function') {
      throw new Error('ChatCredentialCrypto requires atobImpl');
    }
    var saltStr = atobImpl(saltB64);
    var salt = new Uint8Array(saltStr.length);
    for (var i = 0; i < saltStr.length; i += 1) {
      salt[i] = saltStr.charCodeAt(i);
    }
    return salt;
  }

  // PBKDF2 derivation parameters (deriveBits shape, minus salt bytes).
  function pbkdf2Params() {
    return {
      name: 'PBKDF2',
      iterations: PBKDF2_ITERATIONS,
      hash: PBKDF2_HASH,
      bits: PBKDF2_KEY_BITS
    };
  }

  // Authenticator PRF eval input label for the device-wrap key.
  function prfWrapLabel() {
    return PRF_WRAP_LABEL;
  }

  // Authenticator PRF eval input bytes for the device-wrap key.
  function prfEvalBytes(env) {
    var TextEncoderImpl = requireTextEncoder(env);
    return new TextEncoderImpl().encode(PRF_WRAP_LABEL);
  }

  // Derive the Fernet data key from password + server salt.
  async function deriveKeyFromPassword(password, saltB64, env) {
    var subtle = requireSubtle(env);
    var TextEncoderImpl = requireTextEncoder(env);
    var atobImpl = requireFn(env, 'atobImpl');
    var btoaImpl = requireFn(env, 'btoaImpl');
    var enc = new TextEncoderImpl();
    var passwordKey = await subtle.importKey(
      'raw',
      enc.encode(password),
      { name: 'PBKDF2' },
      false,
      ['deriveBits']
    );
    var salt = decodeSaltB64(saltB64, atobImpl);
    var params = pbkdf2Params();
    var derivedBits = await subtle.deriveBits(
      {
        name: 'PBKDF2',
        salt: salt,
        iterations: params.iterations,
        hash: params.hash
      },
      passwordKey,
      params.bits
    );
    return encodeBase64(new Uint8Array(derivedBits), btoaImpl);
  }

  // Generate a device wrap key (AES-GCM 256, non-extractable).
  async function generateWrapKey(env) {
    var subtle = requireSubtle(env);
    return subtle.generateKey(
      { name: 'AES-GCM', length: 256 },
      false,
      ['encrypt', 'decrypt']
    );
  }

  // Probe whether a stored wrap key can still encrypt. False on any
  // failure, including missing capabilities: subtle and randomness resolve
  // inside the probe so a missing provider reports false instead of
  // rejecting.
  async function wrapKeyCanEncrypt(wrapKey, env) {
    try {
      var subtle = requireSubtle(env);
      var getRandomValues = requireFn(env, 'getRandomValues');
      var iv = getRandomValues(new Uint8Array(12));
      await subtle.encrypt({ name: 'AES-GCM', iv: iv }, wrapKey, new Uint8Array([0]));
      return true;
    } catch (_) {
      return false;
    }
  }

  // Wrap a base64 data key under an AES-GCM key.
  async function wrapDataKey(rawKeyB64, aesKey, env) {
    var subtle = requireSubtle(env);
    var getRandomValues = requireFn(env, 'getRandomValues');
    var atobImpl = requireFn(env, 'atobImpl');
    var raw = decodeBase64(rawKeyB64, atobImpl);
    var iv = getRandomValues(new Uint8Array(12));
    var encrypted = await subtle.encrypt({ name: 'AES-GCM', iv: iv }, aesKey, raw);
    return { iv: Array.from(iv), wrapped: Array.from(new Uint8Array(encrypted)) };
  }

  // Unwrap a wrapped record under an AES-GCM key to base64.
  async function unwrapDataKey(record, aesKey, env) {
    var subtle = requireSubtle(env);
    var btoaImpl = requireFn(env, 'btoaImpl');
    var iv = new Uint8Array(record.iv);
    var ciphertext = new Uint8Array(record.wrapped);
    var plain = await subtle.decrypt({ name: 'AES-GCM', iv: iv }, aesKey, ciphertext);
    return encodeBase64(new Uint8Array(plain), btoaImpl);
  }

  // Import authenticator PRF output as an AES-GCM wrap key.
  async function importPrfKey(prfBytes, env) {
    var subtle = requireSubtle(env);
    return subtle.importKey(
      'raw',
      prfBytes,
      { name: 'AES-GCM', length: 256 },
      false,
      ['encrypt', 'decrypt']
    );
  }

  // Whether the given credential constructor supports the PRF extension.
  async function supportsWebAuthnPrf(env) {
    var ctor = env ? env.credentialCtor : undefined;
    if (!ctor) {
      return false;
    }
    if (typeof ctor.getClientCapabilities !== 'function') {
      return false;
    }
    try {
      var caps = await ctor.getClientCapabilities();
      return !!(caps && caps.prf === true);
    } catch (_) {
      return false;
    }
  }

  return {
    PBKDF2_ITERATIONS: PBKDF2_ITERATIONS,
    PBKDF2_HASH: PBKDF2_HASH,
    PBKDF2_KEY_BITS: PBKDF2_KEY_BITS,
    PRF_WRAP_LABEL: PRF_WRAP_LABEL,
    encodeBase64: encodeBase64,
    decodeBase64: decodeBase64,
    decodeSaltB64: decodeSaltB64,
    pbkdf2Params: pbkdf2Params,
    prfWrapLabel: prfWrapLabel,
    prfEvalBytes: prfEvalBytes,
    deriveKeyFromPassword: deriveKeyFromPassword,
    generateWrapKey: generateWrapKey,
    wrapKeyCanEncrypt: wrapKeyCanEncrypt,
    wrapDataKey: wrapDataKey,
    unwrapDataKey: unwrapDataKey,
    importPrfKey: importPrfKey,
    supportsWebAuthnPrf: supportsWebAuthnPrf
  };
}));

File: chatbot-server/tests/credential_crypto.rs
//! Contract for `static/credential-crypto.js`: owned derivation/wrap/PRF
//! algorithms with explicit env dependencies (phase 1, MOD014).
//!
//! The shared unit owns PBKDF2 derivation, AES-GCM wrap-key
//! generation/probe, data-key wrap/unwrap, PRF import/eval bytes plus
//! base64 transforms. Every browser capability arrives via explicit env
//! params; `static/enc-key.js` keeps store lifecycle plus the `EncKey`
//! surface and passes browser globals per call with no inline algorithm
//! copies. Behavioral coverage lives in
//! `fixtures/credential_crypto_test.js` through the stable UMD import
//! against deterministic Node WebCrypto (independent pbkdf2Sync oracle).

use std::path::Path;
use std::process::Command;

fn unit_js() -> &'static str {
    include_str!("../../static/credential-crypto.js")
}

fn enc_js() -> &'static str {
    include_str!("../../static/enc-key.js")
}

#[test]
fn credential_crypto_projects_owned_algorithms() {
    let root = Path::new(env!("CARGO_MANIFEST_DIR")).parent().unwrap();
    let run = Command::new("node")
        .arg(root.join("chatbot-server/tests/fixtures/credential_crypto_test.js"))
        .arg(root.join("static/credential-crypto.js"))
        .output()
        .expect("test image must provide the JS behavior-test runtime");
    assert!(
        run.status.success(),
        "JS credential-crypto behavior: {}",
        String::from_utf8_lossy(&run.stderr)
    );
}

#[test]
fn credential_crypto_js_parses() {
    let source = unit_js();
    let allocator = oxc_allocator::Allocator::default();
    let source_type = oxc_span::SourceType::default()
        .with_module(false)
        .with_jsx(false);
    let ret = oxc_parser::Parser::new(&allocator, source, source_type).parse();
    let errors: Vec<String> = ret.diagnostics.errors().map(ToString::to_string).collect();
    assert!(
        errors.is_empty(),
        "static/credential-crypto.js failed to parse ({} error(s)):\n{}",
        errors.len(),
        errors.join("\n")
    );
}

/// Shared UMD shape owning real algorithms with explicit env deps and no
/// ambient store, DOM, or key-flow reach.
#[test]
fn credential_crypto_is_shared_umd_with_explicit_deps() {
    let unit = unit_js();
    assert!(
        unit.contains("module.exports") && unit.contains("ChatCredentialCrypto"),
        "unit must share the UMD shape (require + browser global)"
    );
    for marker in [
        "deriveKeyFromPassword",
        "generateWrapKey",
        "wrapKeyCanEncrypt",
        "wrapDataKey",
        "unwrapDataKey",
        "importPrfKey",
        "prfEvalBytes",
        "supportsWebAuthnPrf",
        "prfWrapLabel",
        "chatbot-enc-key-wrap-v1",
    ] {
        assert!(
            unit.contains(marker),
            "crypto owner must project {marker}"
        );
    }
    assert!(
        unit.contains("100000") && unit.contains("SHA-256") && unit.contains("AES-GCM"),
        "PBKDF2/AES shapes must stay literal in the owner"
    );
    for forbidden in [
        "crypto.subtle",
        "crypto.getRandomValues",
        "window.",
        "document.",
        "indexedDB",
        "idbGet",
        "navigator",
        "X-Enc-Key",
        "getKeyForRequest",
        "NativeSecureKey",
    ] {
        assert!(
            !unit.contains(forbidden),
            "crypto algorithms must take explicit env instead of reaching {forbidden}"
        );
    }
}

/// enc-key.js wires the crypto owner with explicit browser env and keeps no
/// inline algorithm copies.
#[test]
fn enc_key_wires_crypto_owner_without_inline_copies() {
    let src = enc_js();
    for marker in [
        "ChatCredentialCrypto",
        "cryptoEnv()",
        "deriveKeyFromPassword(password, saltB64, cryptoEnv())",
        "wrapDataKey(rawKeyB64, aesKey, cryptoEnv())",
        "unwrapDataKey(record, aesKey, cryptoEnv())",
        "importPrfKey(prfResults.first, { subtle: crypto.subtle })",
        "prfEvalBytes({ TextEncoderImpl: TextEncoder })",
        "supportsWebAuthnPrf({ credentialCtor: global.PublicKeyCredential })",
    ] {
        assert!(
            src.contains(marker),
            "enc-key.js must wire the crypto owner with explicit env; missing: {marker}"
        );
    }
    for duplicate in [
        "100000",
        "SHA-256",
        "chatbot-enc-key-wrap-v1",
        "AES-GCM",
        "fromCharCode",
        "charCodeAt",
        "deriveBits",
        "importKey",
        "generateKey",
        "0-9a-f",
    ] {
        assert!(
            !src.contains(duplicate),
            "single ownership lives in the crypto/metadata units; enc-key.js must not keep {duplicate}"
        );
    }
    assert!(
        !src.contains("'acct:'") && !src.contains("\"acct:\""),
        "slot prefix lives in the metadata unit"
    );
}
